Defect Dojo/SonarQube integration

twsagarcia opened the pull request:
**Note: DefectDojo is now on Python3 and Django 2.2.1 Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

When submitting a pull request, please make sure you have completed the following checklist:

- [x] Your code is flake8 compliant 
- [x] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [x] Model changes must include the necessary migrations in the dojo/dd_migrations folder.
- [x] Add applicable tests to the unit tests.

twsagarcia commented:
This feature implements two kind of integrations with SonarQube. 

- from SonarQube issues to DefectDojo findings:
This is the usual DefectDojo workflow. Just select Scan type _SonarQube Scan_ on _Import Scan Results_ page. As this feature goes directly to SonarQube server to gather the open issues, there is no need to upload a file.
Take in consideration that SonarQube Project name must match DefectDojo Product name to work fine.
This have been covered from GUI, API v1 and API v2.

- DefectDojo finding update to SonarQube issue:
This happens when a DefectDojo finding is updated. Finding current status is mapped to a valid SonarQube status/resolution and SonarQube issue transitions to this status/resolution.
As original SonarQube issue is updated, if the finding in DefectDojo is updated to mitigated/false positive or risk accepted, future import scans won't import the issue.
The finding must have been imported by SonarQube scan to work.

For both functionalities, it requires:
- to configure a tool type beforehand called "SonarQube".
- to configure a tool from Tool Type "SonarQube". Requires:
  * Url
  * Authentication details: supports API Key or Username/Password.

twsagarcia commented:
@rjimgal @aaronweaver @agigleux @jpescalona @AdrienGuillerme @praveendvd @ptrovatelli 
I added you to this review as you seemed interested on having a working SonarQube Integration. #810 

Could you review and provide some feedback? Thanks a lot.

praveendvd commented:
@twsagarcia  

I am getting the below error, i am running a local host instance of both defect dojo and sonarQube

The git commit version is :  544c7b3: Renumbered migration to 0011 after latest dev merge [2019-08-05 11:14:00 +0200]

Could you guide me on where i am going wrong.

![image](https://user-images.githubusercontent.com/45095911/62771472-3692e200-ba95-11e9-825d-2c071ebf63ad.png)

ptrovatelli commented:
i'll have a look asap. I like the idea which differs from what is done everywhere else in the app although on my part i'm looking for a similar process for a set of several security reports.

ptrovatelli commented:
@praveendvd can you share your tool configuration? is your sonar on https? @twsagarcia have you tested with an https sonar? 

I have used: 
- name: sonarqube
- url: the sonar url including until /api 
- tool type: SonarQube
- authentication type: Api Key
- API key: my api key. 

You can also try "docker-compose build" before anything else in case your images are off.

ptrovatelli commented:
@twsagarcia thank you for your work, it's a great idea! 
here is my first feeback:
- There is already a "SonarQube Scan" existing, based on html. I think both solutions can co-exist so that users can choose the one that best fits their needs. 
Can you rename your importer to maybe "SonarQubeDirect" (everywhere, all fields, variable names etc so that we know what we're looking at is specific use case added for this importer? 
- What sonar version are you using? i'm on 6.5 and I think I may be facing an issue with this. 
It works when there is no findings, but as soon as there are findings i'm getting error: 
```
Internal Server Error: /engagement/12/import_scan_results
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.5/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "./dojo/engagement/views.py", line 523, in import_scan_results
    parser = import_parser_factory(file, t, active, verified)
  File "./dojo/tools/factory.py", line 146, in import_parser_factory
    parser = SonarQubeApiImporter(test)
  File "./dojo/tools/sonarqube/importer.py", line 22, in __init__
    self.items = self.import_issues(self.client, test)
  File "./dojo/tools/sonarqube/importer.py", line 62, in import_issues
    from_hotspot = security_issue['fromHotspot']
KeyError: 'fromHotspot'
```

Looking at the `security_issue` content, there is indeed no `fromHotspot` key. 

Thanks!

praveendvd commented:
> @praveendvd can you share your tool configuration? is your sonar on https? @twsagarcia have you tested with an https sonar?
> 
> I have used:
> 
> * name: sonarqube
> * url: the sonar url including until /api
> * tool type: SonarQube
> * authentication type: Api Key
> * API key: my api key.
> 
> You can also try "docker-compose build" before anything else in case your images are off.

Hi @ptrovatelli Thank you so much, but  I do build new images every time. I stop all containers, deletes all images and volumes before building the image.

Below is my configuration details, I am running both docker and sonar as localhost instance.

![image](https://user-images.githubusercontent.com/45095911/62823556-fc573c80-bb89-11e9-998b-de2f6f9e9981.png)

Please do let me know, where I am going wrong thank you so much. Below is my commit I am on

544c7b3: Renumbered migration to 0011 after latest dev merge [2019-08-05 11:14:00 +0200]

ptrovatelli commented:
@praveendvd you have put your api key in the "apî key title" field, have you put it in the "api key" field too?

praveendvd commented:
> @praveendvd you have put your api key in the "apî key title" field, have you put it in the "api key" field too?

@ptrovatelli  Yes i have given in both fields. I also tried with user name and password

ptrovatelli commented:
@praveendvd url fields needs the API url; probably localhost:9000/api in your case. 
if still not working, you can post the uwsgi container logs, though we probably should continue this discussion on slack!

praveendvd commented:
> @praveendvd url fields needs the API url; probably localhost:9000/api in your case.
> if still not working, you can post the uwsgi container logs, though we probably should continue this discussion on slack!

@ptrovatelli  please find the log below:

[text.txt.zip](https://github.com/DefectDojo/django-DefectDojo/files/3489170/text.txt.zip)

ptrovatelli commented:
@praveendvd I have opened a thread on slack. at this point I believe this is probably a configuration issue, we'll post back here if an issue is confirmed

ptrovatelli commented:
@twsagarcia one more thing: it may be a bit tricky to work on project name in sonar as those are not unique. Alghough it's less obvious when creating the product in defectDojo, it may be safer to work on the `project key` for identifying the project we're working on
This is an API call based on the project key: 
`https://sonarcloud.io/dashboard?id=tomcat_180410`

twsagarcia commented:
Hi all, thank you for your support.

I'll try to to reply here and will update the first post with all the info when possible.

First post update:
- Added a dummy configuration that should work.
- Added SonarQube version tested.

@ptrovatelli If there is no Security Hotspot support, I guess you are using an old SonarQube version. I could try to make your version compatible but I can't test it properly. I will review it and post a fix asap.

@ptrovatelli According to using old "SonarQube Scan" or a new "SonarQubeDirect"... I thought "SonarQube Scan" was deprecated... In fact, I found it commented everywhere. I'd like to minimize confusions here but not sure if having 2 kind of SonarQube scans is going to be help with that. Let's continue the discussion in Slack.

@praveendvd Please take a look at the configuration example I posted in the first post.

@praveendvd I replied to you in Slack. Let's continue the discussion there if needed.

twsagarcia commented:
Solved the conflicts so far but build is failing due to a migration file is missing.... This has been caused by a recent merge to dev. Previous builds were fine.

I think the commit that caused the issue is:
Django 2.2.1 Upgrade
74fc149b Maffooch <codymaffucci@codys-macbook-pro.local> on 6/6/19 at 9:18 PM

@Maffooch could you review?

ptrovatelli commented:
@twsagarcia about project name vs project key: it may be a big constraint to make ppl name their product based on the sonar project key. for now maybe we could just raise an error when several projects are found based on the project name? it can be told easily from the json response of that `projects/search?` query.
another way to deal with this would be to add an override at product level configuration in defectDojo to give the sonar project name. 
another thought: not blocking right now, but i think there may be setups where serveral applications, possibly with different sonar url can live in the same defectDojo instance. It would be nice to have a configuration override at product level again, in order to give the sonar url and api key .

ptrovatelli commented:
@twsagarcia  the previous sonarQube import was commented out because the export code that produced the html code for defectDojo was closed source so few people could use the import. On may 2019 we have opened that export code so that everyone can use it. Then what needed to be done was just uncomment the code, make it work for python3 plus a couple of corrections/ improvments that I have done at about the same time as your PR . 
the full history can be found here: https://github.com/DefectDojo/django-DefectDojo/issues/810
Right now I cannot tell if your method will work for all cases which is why I would like to keep both solutions; even if everything else work, we can imagine there would setups where defectDojo cannot communicate with sonar because one of them is in a DMZ for example, or security constraints forbids sharing sonar api key; there could also be other incompatibility like the one I had with "hotspot"; still we could import reports with the old method

ptrovatelli commented:
Hi @twsagarcia; have you been able to link defectDojo to a sonarcloud.io instance? the call https://sonarcloud.io/api/projects/search?q=tomcat requires authentication: 
`{"errors":[{"msg":"Insufficient privileges"}]}
`
I don't have an account, not sure if we can create an token with a free account and access public projects with it. 
The data is publicly accessible through, accessible here in my example https://sonarcloud.io/dashboard?id=tomcat_180410

This was possible with the html export, as all calls done don't require authentication. 
You can check the api calls made here: 
https://github.com/soprasteria/sonar-report/blob/master/index.js

Other than that, with a token I have been able to access my own sonarqube instance. didn't have time to check results in details though.

twsagarcia commented:
Hi @ptrovatelli,

I'm happy that the hotspot patch worked fine.

Sorry but I haven't tested this on SonarCloud.io. It seems to use SonarQube API so should be compatible. 
I just created an account on sonarcloud.io and a free account can generate a token without issues. However user/password is also available and supported by this integration.

Unfortunately, I haven't got a lot of time to deep into the sonar-report tool and SonarCloud.io.

About the open previous suggestions/comments:

- Ok with keeping both SonarQube integrations together. To avoid conflicts, I will call this new scan type "SonarQube API Import". I will review the code to make this name consistent in the code.

- About project name vs project key: I think I will add some additional fields in the product section: SonarQube Product Key and SonarQube Tool (as every Tool already have Sonar Url + Auth details)...  If the product don't provide them, current behavior would remain.

Hopefully, someone else will review this PR in the next few days and hopefully I will have time to undertake your suggested changes.

ptrovatelli commented:
@twsagarcia if you do that it'll be great, there isn't much more.
thanks!

praveendvd commented:
> @twsagarcia if you do that it'll be great, there isn't much more.
> thanks!

Hi, i am getting connection refused error even for jira local server instance #1487 . So does there any issue with supporting local server instance in Defectdojo as i can see even sonarqube being tested here is cloud version @ptrovatelli

ptrovatelli commented:
@praveendvd what IP do you use to reach jira/sonar from defectdojo? From a network point of view, the container and the host are two distinct servers so "localhost" won't work. have you tried  the link I posted in slack that gave solutions in order to reach the host from the container?

praveendvd commented:
Thank you so much, I missed that out sorry for that

bilalk88 commented:
> @praveendvd what IP do you use to reach jira/sonar from defectdojo? From a network point of view, the container and the host are two distinct servers so "localhost" won't work. have you tried the link I posted in slack that gave solutions in order to reach the host from the container?

Hi @ptrovatelli 
Can you please help me configuring jira integration with defectdojo. As I'm getting an error after entering the proper details for JIRA.
Error : "Unable to authenticate. Please check the URL, username, and password."

Console log : 
Starting new HTTPS connection (1): JIRA-URL.com:443 HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc22b10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)) while doing GET https://JIRA-URL.com/rest/api/2/serverInfo [{u'headers': {'Accept-Encoding': 'gzip, deflate', u'Accept': u'application/json,*.*;q=0.9', 'User-Agent': 'python-requests/2.22.0', 'Connection': 'keep-alive', u'X-Atlassian-Token': u'no-check', u'Cache-Control': u'no-cache', u'Content-Type': u'application/json'}, 'params': None}] Got ConnectionError [HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc22b10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))] errno:None on GET https://JIRA-URL.com/rest/api/2/serverInfo {'request': <PreparedRequest [GET]>, 'response': None}\{'request': <PreparedRequest [GET]>, 'response': None} Got recoverable error from GET https://JIRA-URL.com/rest/api/2/serverInfo, will retry [1/3] in 13.057812708s. Err: HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc22b10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)) Starting new HTTPS connection (2): JIRA-URL.com:443 HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc22d10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)) while doing GET https://JIRA-URL.com/rest/api/2/serverInfo [{u'headers': {'Accept-Encoding': 'gzip, deflate', u'Accept': u'application/json,*.*;q=0.9', 'User-Agent': 'python-requests/2.22.0', 'Connection': 'keep-alive', u'X-Atlassian-Token': u'no-check', u'Cache-Control': u'no-cache', u'Content-Type': u'application/json'}, 'params': None}] Got ConnectionError [HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc22d10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))] errno:None on GET https://JIRA-URL.com/rest/api/2/serverInfo {'request': <PreparedRequest [GET]>, 'response': None}\{'request': <PreparedRequest [GET]>, 'response': None} Got recoverable error from GET https://JIRA-URL.com/rest/api/2/serverInfo, will retry [2/3] in 28.5262903487s. Err: HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc22d10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)) Starting new HTTPS connection (3): JIRA-URL.com:443 HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc42d90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)) while doing GET https://JIRA-URL.com/rest/api/2/serverInfo [{u'headers': {'Accept-Encoding': 'gzip, deflate', u'Accept': u'application/json,*.*;q=0.9', 'User-Agent': 'python-requests/2.22.0', 'Connection': 'keep-alive', u'X-Atlassian-Token': u'no-check', u'Cache-Control': u'no-cache', u'Content-Type': u'application/json'}, 'params': None}] Got ConnectionError [HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc42d90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))] errno:None on GET https://JIRA-URL.com/rest/api/2/serverInfo {'request': <PreparedRequest [GET]>, 'response': None}\{'request': <PreparedRequest [GET]>, 'response': None} Got recoverable error from GET https://JIRA-URL.com/rest/api/2/serverInfo, will retry [3/3] in 16.6726615373s. Err: HTTPSConnectionPool(host='JIRA-URL.com', port=443): Max retries exceeded with url: /rest/api/2/serverInfo (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f50fdc42d90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',)) 

I'm using docker based DefectDojo.

praveendvd commented:
Are you using defectdojo docker ?

bilalk88 commented:
> Are you using defectdojo docker ?

yes. I'm using defectdojo docker.

praveendvd commented:
Is the your Jira available publicly , try pinging the Jira from container. 

Use docker exec -it ,

To interact with the container

bilalk88 commented:
> Is the your Jira available publicly , try pinging the Jira from container.
> 
> Use docker exec -it ,
> 
> To interact with the container

I tried that but none of the container having ping command. As getting an error "command not found". I also tried to get inside the container and tried to perform apt-get update but also getting the same error "Temporary failure resolving 'deb.debian.org'".

Below is the log content
docker exec -it --user root django-defectdojo_celeryworker_1 /bin/bash
root@bfb79b55fd3f:/app# apt-get update
Err:1 http://security.debian.org/debian-security stretch/updates InRelease
  Temporary failure resolving 'security.debian.org'
Err:2 http://deb.debian.org/debian stretch InRelease
  Temporary failure resolving 'deb.debian.org'
Err:3 http://deb.debian.org/debian stretch-updates InRelease
  Temporary failure resolving 'deb.debian.org'
Reading package lists... Done
W: Failed to fetch http://deb.debian.org/debian/dists/stretch/InRelease  Temporary failure resolving 'deb.debian.org'
W: Failed to fetch http://security.debian.org/debian-security/dists/stretch/updates/InRelease  Temporary failure resolving 'security.debian.org'
W: Failed to fetch http://deb.debian.org/debian/dists/stretch-updates/InRelease  Temporary failure resolving 'deb.debian.org'
W: Some index files failed to download. They have been ignored, or old ones used instead.

Also when I try to check the /etc/resolv.conf file found below content : 
root@bfb79b55fd3f:/app# cat /etc/resolv.conf
nameserver 127.0.0.11
options rotate timeout:1 ndots:0
root@bfb79b55fd3f:/app#

praveendvd commented:
Just restart you docker service , and try

bilalk88 commented:
> Just restart you docker service , and try
> 
> Sudo service docker restart ,
> 
> The current issue is dns related , as it cannot resolve even Debian sites

tried that too. For few minutes it's able to resolve it but after sometime let say 10 15 minutes again it's not able to resolve it.
Also I followed the steps mentioned in the link https://askubuntu.com/questions/881843/dns-issue-with-docker-image
but still no luck

ptrovatelli commented:
@bilalk88  @praveendvd can we please continue this discussion in slack and keep this thread just for exchanges about the sonarqube integration code review? thanks! :)

twsagarcia commented:
@ptrovatelli All proposed changes were implemented on the last commit.

@Maffooch To ease your verification work, I have tested the implementation on [SonarCloud.io](https://sonarcloud.io) (SonarQube on the cloud) and it's working fine. You can create a free account and use your username/password to get findings of any public cloud project. Otherwise it's also possible to generate a token for your account and use the token to authenticate.
Let me know at Slack if you need assistance.

ptrovatelli commented:
@twsagarcia sorry but more migrations have been added again; can you start at index 17 please? 
https://github.com/DefectDojo/django-DefectDojo/tree/dev/dojo/db_migrations

if the previous sonar import is back online, we'll try to merge quickly :)

ptrovatelli commented:
@twsagarcia something's off with travis. the initializer container is not up to date which is why you still have 2 number 15. Hold on i'll have a look. It used to be done in build_containers function in travis.sh, I don't know why it was removed a couple of month ago.

twsagarcia commented:
@ptrovatelli Don't worry. My last merge from upstream/dev in branch SonarQube had two 00015 migrations... Although was solved a bit later my branch was broken until I merged from upstream/dev again. Anyway I've renamed my migration file again and Travis build is ok now.

ptrovatelli commented:
@twsagarcia there is a little problem with the dojo_sonarqube_issue.key primary key. i was very lucky to find this with my first test
i'm still working on the tomcat project (sonarqube project key: tomcat_180410 ) 
the import fails with error: 
```
IntegrityError at /engagement/5/import_scan_results
(1062, "Duplicate entry 'AWK40IGV-pl6AHs22MdB' for key 'sonarqube_issue_id'")
Request Method:	POST
Request URL:	http://localhost:8081/engagement/5/import_scan_results
Django Version:	2.2.1
Exception Type:	IntegrityError
Exception Value:	
(1062, "Duplicate entry 'AWK40IGV-pl6AHs22MdB' for key 'sonarqube_issue_id'")
Exception Location:	/usr/local/lib/python3.5/site-packages/MySQLdb/connections.py in query, line 224
Python Executable:	/usr/local/bin/uwsgi
Python Version:	3.5.7
Python Path:	
['/usr/local/lib/python3.5/site-packages/git/ext/gitdb',
 '/app',
 '.',
 '',
 '/usr/local/lib/python35.zip',
 '/usr/local/lib/python3.5',
 '/usr/local/lib/python3.5/plat-linux',
 '/usr/local/lib/python3.5/lib-dynload',
 '/usr/local/lib/python3.5/site-packages',
 '/usr/local/lib/python3.5/site-packages/gitdb/ext/smmap']
Server time:	Tue, 27 Aug 2019 11:56:54 +0000
```

After a bit of analysis, i've found out that this is due to a case insensitive comparison! 
Both keys are found in the output of: 
`curl "https://sonarcloud.io/api/issues/search?p=3&componentKeys=tomcat_180410&types=BUG%2CVULNERABILITY%2CCODE_SMELL" > issue_3.json`

they are very different vulns but have the same key when looking at it without case sensitivity: 
`"key": "AWK40IGV-pl6AHs22MdB",`
vs
`"key": "AWK40IGV-pl6AHs22Mdb",`

By the way i'm glad we still have the html import :) . with the api importer, it's not possible to import twice the same vuln. I need to be able to do this so that I can compare list of vuln in engagements to see what vuln was added, fixed, or is still present. 
Thanks!

twsagarcia commented:
@ptrovatelli Really good catch. 
Just to clarify the issue, this error is related to current MySQL DB configuration which is performing a case insensitive comparison for searches and indexes... Django (which do all string equality comparisons using case sensitive) warns specifically about this issue when using MySQL with default collation (case insensitive)
https://docs.djangoproject.com/en/2.2/ref/databases/#collation-settings

As Django doesn’t provide a way to set this on the model definition, I fixed it manually in the migration forcing to update the table collation just after creating the table.

Having this also covered, I think that's all. 
However Travis seems to be failing now :(

ptrovatelli commented:
good idea on keeping only vulnerabilities; it's much, much faster now! 
I have compared the results between the html and the api importer on the tomcat sonarcloud project: they both import 239 findings. It's all good to me thanks :)

ptrovatelli commented:
@Maffooch what do you think?

Maffooch commented:
@ptrovatelli Have not gotten a lot of dojo time since school resumed. Hope to grab a bit this weekend. I trust your judgement though.
